homeassistant/components/vicare/number.py

```diff
@@ -0,0 +1,166 @@
+"""Number for ViCare."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from contextlib import suppress
+from dataclasses import dataclass
+import logging
+from typing import Any
+
+from PyViCare.PyViCareDevice import Device as PyViCareDevice
+from PyViCare.PyViCareDeviceConfig import PyViCareDeviceConfig
+from PyViCare.PyViCareHeatingDevice import (
+    HeatingDeviceWithComponent as PyViCareHeatingDeviceWithComponent,
+)
+from PyViCare.PyViCareUtils import (
+    PyViCareInvalidDataError,
+    PyViCareNotSupportedFeatureError,
+    PyViCareRateLimitError,
+)
+from requests.exceptions import ConnectionError as RequestConnectionError
+
+from homeassistant.components.number import NumberEntity, NumberEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory, UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ViCareRequiredKeysMixin
+from .const import DOMAIN, VICARE_API, VICARE_DEVICE_CONFIG
+from .entity import ViCareEntity
+from .utils import is_supported
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class ViCareNumberEntityDescription(NumberEntityDescription, ViCareRequiredKeysMixin):
+    """Describes ViCare number entity."""
+
+    value_setter: Callable[[PyViCareDevice, float], Any] | None = None
+
+
+CIRCUIT_ENTITY_DESCRIPTIONS: tuple[ViCareNumberEntityDescription, ...] = (
+    ViCareNumberEntityDescription(
+        key="heating curve shift",
+        name="Heating curve shift",
+        icon="mdi:plus-minus-variant",
+        entity_category=EntityCategory.CONFIG,
+        value_getter=lambda api: api.getHeatingCurveShift(),
+        value_setter=lambda api, shift: api.setHeatingCurve(
+            shift, api.getHeatingCurveSlope()
+        ),
```

Comment:
Is there a way to avoid the get before setting?

Reply:
I had no idea how. The api exposes only the ability to change both at the same time.

Reply:
I think that's fine. Workarounds based on the states or similar would be too error-prone, I think.

Reply:
Ah, just one small thing I learned lately: use parenthesis for multi-line lambdas

Reply:
Like this? 

```suggestion
        value_setter=lambda api, shift: (api.setHeatingCurve(
            shift, api.getHeatingCurveSlope()
        )),
```

Reply:
Nearly, when running `black` it looks like this, which is more readable: 

```suggestion
        value_setter=lambda api, shift: (
            api.setHeatingCurve(shift, api.getHeatingCurveSlope())
        ),
```